k9mail/src/main/java/com/fsck/k9/activity/compose/AttachmentPresenter.java

```diff
@@ -0,0 +1,371 @@
+package com.fsck.k9.activity.compose;
+
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+
+import android.annotation.TargetApi;
+import android.app.Activity;
+import android.app.LoaderManager;
+import android.content.ClipData;
+import android.content.Context;
+import android.content.Intent;
+import android.content.Loader;
+import android.net.Uri;
+import android.os.Build;
+import android.os.Bundle;
+import android.os.Handler;
+
+import com.fsck.k9.activity.compose.ComposeCryptoStatus.AttachErrorState;
+import com.fsck.k9.activity.loader.AttachmentContentLoader;
+import com.fsck.k9.activity.loader.AttachmentInfoLoader;
+import com.fsck.k9.activity.misc.Attachment;
+import com.fsck.k9.activity.misc.Attachment.LoadingState;
+import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.Multipart;
+import com.fsck.k9.mail.Part;
+import com.fsck.k9.mail.internet.MimeUtility;
+import com.fsck.k9.mailstore.LocalBodyPart;
+import com.fsck.k9.provider.AttachmentProvider;
+
+
+public class AttachmentPresenter {
+    private static final String STATE_KEY_ATTACHMENTS = "com.fsck.k9.activity.MessageCompose.attachments";
+    private static final String STATE_KEY_WAITING_FOR_ATTACHMENTS = "waitingForAttachments";
+    private static final String STATE_KEY_NEXT_LOADER_ID = "nextLoaderId";
+
+    private static final String LOADER_ARG_ATTACHMENT = "attachment";
+    private static final int LOADER_ID_MASK = 1 << 6;
+    public static final int MAX_TOTAL_LOADERS = LOADER_ID_MASK -1;
+    public static final int REQUEST_CODE_ATTACHMENT_URI = 1;
+
+
+    // injected state
+    private final Context context;
+    private final AttachmentMvpView attachmentMvpView;
+    private final LoaderManager loaderManager;
+
+    // persistent state
+    private LinkedHashMap<Uri, Attachment> attachments;
```

Comment:
attachments are unique by their source uri, thus identified by them. Using a Map here to avoid a need for `indexOf` in `onLoadFinished` operations below since we replace the Attachment objects with new state (though they aren't immutable yet). LinkedHashMap to preserve order since this is the principal list of attachments from which the view is restored in `onRestoreInstanceState`